Sources/GRPCCodeGen/Internal/Translator/ClientCodeTranslator.swift

```diff
@@ -0,0 +1,414 @@
+        .generic(
+          wrapper: .member("MessageSerializer"),
+          wrapped: .member(
+            self.methodInputOutputTypealias(for: method, service: service, type: .input)
+          )
+        )
+      )
+    )
+  }
+
+  private func deserializerParameter(
+    for method: CodeGenerationRequest.ServiceDescriptor.MethodDescriptor,
+    in service: CodeGenerationRequest.ServiceDescriptor
+  ) -> ParameterDescription {
+    return ParameterDescription(
+      label: "deserializer",
+      type: ExistingTypeDescription.some(
+        .generic(
+          wrapper: .member("MessageDeserializer"),
+          wrapped: .member(
+            self.methodInputOutputTypealias(for: method, service: service, type: .output)
+          )
+        )
+      )
+    )
+  }
+
+  private func bodyParameter(
+    for method: CodeGenerationRequest.ServiceDescriptor.MethodDescriptor,
+    in service: CodeGenerationRequest.ServiceDescriptor
+  ) -> ParameterDescription {
+    let clientStreaming = method.isOutputStreaming ? "Stream" : "Single"
+    let closureParameterType = ExistingTypeDescription.generic(
+      wrapper: .member(["ClientResponse", clientStreaming]),
+      wrapped: .member(
+        self.methodInputOutputTypealias(for: method, service: service, type: .output)
+      )
+    )
+
+    let bodyClosure = ClosureSignatureDescription(
+      parameters: [.init(type: closureParameterType)],
+      keywords: [.async, .throws],
+      returnType: .identifierType(.member("R")),
+      sendable: true,
+      escaping: true
+    )
+    return ParameterDescription(name: "body", closureDescription: bodyClosure)
+  }
+
+  private func makeClientStruct(
+    for service: CodeGenerationRequest.ServiceDescriptor,
+    in codeGenerationRequest: CodeGenerationRequest
+  ) -> Declaration {
+    let clientProperty = Declaration.variable(
+      kind: .let,
+      left: "client",
+      type: .member(["GRPCCore", "GRPCClient"])
+    )
+    let initializer = self.makeClientVariable()
+    let methods = service.methods.compactMap {
+      self.makeClientMethod(for: $0, in: service, from: codeGenerationRequest)
+    }
+
+    return .struct(
+      StructDescription(
+        name: "\(service.namespacedPrefix)Client",
+        conformances: ["\(service.namespacedTypealiasPrefix).ClientProtocol"],
+        members: [clientProperty, initializer] + methods
+      )
+    )
+  }
+
+  private func makeClientVariable() -> Declaration {
+    let initializerBody = Expression.assignment(
+      left: .memberAccess(
+        MemberAccessDescription(left: .identifierPattern("self"), right: "client")
+      ),
+      right: .identifierPattern("client")
+    )
+    return .function(
+      signature: .init(
+        kind: .initializer,
+        parameters: [.init(label: "client", type: .member(["GRPCCore", "GRPCClient"]))]
+      ),
+      body: [CodeBlock(item: .expression(initializerBody))]
+    )
+  }
+
+  private func getGRPCMethodName(
```

Comment:
"grpc method name" is misleading here, it sounds like the name of the RPC, `clientMethod` is a bit clearer. Also we should just pass in two parameters here: input and output streaming